---
title: "GSB_Microarry"
output: html_document
date: "2024-08-22"
---

```{r setup, include=FALSE}
#for the aalysis from .CEL files
getwd()
suppressPackageStartupMessages({library("maEndToEnd")})
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#libraries required 

#General Bioconductor packages
    library(Biobase)
    library(oligoClasses)
     
#Annotation and data import packages
    library(ArrayExpress)
    library(pd.hugene.1.0.st.v1)
    library(hugene10sttranscriptcluster.db)
     
#Quality control and pre-processing packages
    library(oligo)
    library(arrayQualityMetrics)
     
#Analysis and statistics packages
    library(limma)
    library(topGO)
    library(ReactomePA)
    library(clusterProfiler)
     
#Plotting and color options packages
    library(gplots)
    library(ggplot2)
    library(geneplotter)
    library(RColorBrewer)
    library(pheatmap)
    library(enrichplot)
     
#Formatting/documentation packages
   #library(rmarkdown)
   #library(BiocStyle)
    library(dplyr)
    library(tidyr)

#Helpers:
    library(stringr)
    library(matrixStats)
    library(genefilter)
    library(openxlsx)
   #library(devtools)
    library(affy)
```


```{r}
#defines directory in library/R

#raw_data_dir <- tools::R_user_dir(package = "maEndToEnd", which = "cache")

#if (!dir.exists(raw_data_dir)) {
#    dir.create(raw_data_dir, recursive = TRUE)
#}

working_dir <- getwd()

# Define the directory path within the working directory
raw_data_dir <- file.path(working_dir, "maEndToEnd1")

# Check if the directory exists, and create it if it doesn't
if (!dir.exists(raw_data_dir)) {
    dir.create(raw_data_dir, recursive = TRUE)
}
```

#The first step of the analysis is to download the raw data CEL files. These files are produced by the array scanner software and contain the measured probe intensities.

# use the getAEfunction from the ArrayExpress Bioconductor package to obtain the ftp links to the raw data files 

```{r}

options(timeout = 1200)  # Set timeout to 600 seconds

anno_AE <- getAE("E-MTAB-6940", path = raw_data_dir, type = "raw")
download.file(url = "https://www.ebi.ac.uk/biostudies/files/E-MTAB-6940/E-MTAB-6940.sdrf.txt",
              destfile = file.path(raw_data_dir, "E-MTAB-6940.sdrf.txt"), 
              mode = "wb")

```


```{r}
sdrf_location <- file.path(raw_data_dir, "E-MTAB-6940.sdrf.txt")
SDRF <- read.delim(sdrf_location)

rownames(SDRF) <- SDRF$Array.Data.File
SDRF <- AnnotatedDataFrame(SDRF)
```


```{r}
#This automatically creates an ExpressionSet, fills the sections “array data” with the data from the CEL files and uses the correct chip annotation package, in this case pd.hugene.1.0.st.v1 (the chip-type is also stored in the .CEL files).

raw_data <- oligo::read.celfiles(filenames = file.path(raw_data_dir, 
                                                SDRF$Array.Data.File),
                                    verbose = FALSE, phenoData = SDRF)
stopifnot(validObject(raw_data))

```
```{r}

#The pData function of the Biobase package directly accesses the phenoData in the ExpressionSet raw_data. With the head() function, we can view the first six lines of the table. 

head(Biobase::pData(raw_data))

```
```{r}
#The columns of interest for us are the following: identifiers of the individuals, i.e. columns “Source.Name”, “Characteristics.individual.” disease of the individual, i.e. “Factor.Value.disease.” Cell type, i.e. “Factor.Value.cell.type.”

Biobase::pData(raw_data) <- Biobase::pData(raw_data)[, c("Source.Name",
                                     "Characteristics.individual.",
                                     "Factor.Value.disease.",
                                     "Factor.Value.cell.type.")]

```


```{r}

#The first step after the initial data import is the quality control of the data. Here we check for outliers and try to see whether the data clusters as expected, e.g. by the experimental conditions

Biobase::exprs(raw_data)[1:5, 1:5]

#The rows represent the microarray probes, i.e.  the single DNA locations on the chip, while the columns represent one microarray,

exp_raw <- log2(Biobase::exprs(raw_data))
PCA_raw <- prcomp(t(exp_raw), scale. = FALSE)

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    Disease = pData(raw_data)$Factor.Value.disease.,
                    Phenotype = pData(raw_data)$Factor.Value.cell.type.,
                    Individual = pData(raw_data)$Characteristics.individual.)

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = Disease, colour = Phenotype)) +
  ggtitle("PCA plot of the log-transformed raw expression data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5))#+
  #coord_fixed(ratio = sd_ratio) #+
  #scale_shape_manual(values = c(4,15)) + 
  #scale_color_manual(values = c("darkorange2", "dodgerblue4"))

#The PCA plot (performed on the log-intensity scale) of the raw data shows that the first principal component doesn not differentiates much between the cell types This means that the disease type is not a major driver of gene expression differences.

oligo::boxplot(raw_data, target = "core", 
               main = "Boxplot of log2-intensitites for the raw data")

```

```{r}

arrayQualityMetrics(expressionset = raw_data,
    outdir = tempdir(),
    force = TRUE, do.logtransform = TRUE,
    intgroup = c("Factor.Value.disease.", "Factor.Value.cell.type."))

```
```{r}
#The package oligo allows us to perform background correction, normalization and summarization in one single step using a deconvolution method for background correction, quantile normalization and the RMA (robust multichip average) algorithm for summarization.

palmieri_eset <- oligo::rma(raw_data, target = "core", normalize = FALSE)
```

```{r}
row_medians_assayData <- 
  Biobase::rowMedians(as.matrix(Biobase::exprs(palmieri_eset)))

RLE_data <- sweep(Biobase::exprs(palmieri_eset), 1, row_medians_assayData)

RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- 
  tidyr::gather(RLE_data, patient_array, log2_expression_deviation)

ggplot2::ggplot(RLE_data_gathered, aes(patient_array,
                                       log2_expression_deviation)) + 
  geom_boxplot(outlier.shape = NA) + 
  coord_cartesian(ylim = c(-2, 2)) + 
  theme(axis.text.x = element_text(colour = "aquamarine4", 
                                  angle = 60, size = 6.5, hjust = 1 ,
                                  face = "bold"))

#Note that the y-axis now displays for each microarray the deviation of expression intensity from the median expression of the respective single transcripts across arrays. Boxes with a larger extension therefore indicate an unusually high deviation from the median in a lot of transcripts, suggesting that these arrays are different from most of the others in some way.
```

```{r}

palmieri_eset_norm <- oligo::rma(raw_data, target = "core")
```

```{r}
exp_palmieri <- Biobase::exprs(palmieri_eset_norm)
PCA <- prcomp(t(exp_palmieri), scale = FALSE)

percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Disease = 
                     Biobase::pData(palmieri_eset_norm)$Factor.Value.disease.,
                    Phenotype = 
                     Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.)


ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = Disease, colour = Phenotype)) +
  ggtitle("PCA plot of the calibrated, summarized data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) #+
  #coord_fixed(ratio = sd_ratio) +
  #scale_shape_manual(values = c(4,15)) + 
  #scale_color_manual(values = c("darkorange2", "dodgerblue4"))

```

```{r}
#phenotype_names <- ifelse(str_detect(pData
 #                                   (palmieri_eset_norm)$Factor.Value.cell.type.,
  #                           "CCSC"), "CTC.", "mCCSC.")

library(dplyr)

phenotype_names <- case_when(
  str_detect(pData(palmieri_eset_norm)$Factor.Value.cell.type., "mCCSC") ~ "mCCSC.",
  str_detect(pData(palmieri_eset_norm)$Factor.Value.cell.type., "CTC") ~ "CTC.",
  TRUE ~ "CCSC."  # Default case
)

disease_names <- ifelse(str_detect(pData
                                    (palmieri_eset_norm)$Factor.Value.disease.,
                             "metastatic colorectal cancer"), "mCC", "CC")

annotation_for_heatmap <- 
  data.frame(Phenotype = phenotype_names,  Disease = disease_names)

row.names(annotation_for_heatmap) <- row.names(pData(palmieri_eset_norm))



dists <- as.matrix(dist(t(exp_palmieri), method = "manhattan"))

rownames(dists) <- row.names(pData(palmieri_eset_norm))
hmcol <- rev(colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255))
colnames(dists) <- NULL
diag(dists) <- NA

ann_colors <- list(
  Phenotype = c(mCCSC. = "chartreuse4", CTC. = "burlywood3", CCSC. = "darkorange2"),
  Disease = c(CC = "blue4", mCC = "cadetblue2")
                   )
pheatmap(dists, col = (hmcol), 
         annotation_row = annotation_for_heatmap,
         annotation_colors = ann_colors,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                         max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering heatmap for the calibrated samples")

```

```{r}
#We now filter out lowly expressed genes. Microarray data commonly show a large numberof probes in the background intensity range. These probes also do not change much across arrays. Hence they combine a low variance with a low intensity. Thus, they could end up being detected as differentially expressed although they are barely above the “detection” limit and are not very informative in general.


palmieri_medians <- rowMedians(Biobase::exprs(palmieri_eset_norm))

hist_res <- hist(palmieri_medians, 100, col = "cornsilk1", freq = FALSE, 
            main = "Histogram of the median intensities", 
            border = "antiquewhite4",
            xlab = "Median intensities")

```

```{r}
#In the histogram from previous chunkof the gene-wise medians, we can clearly see an enrichment of low medians on the left hand side. These represent the genes we want to filter. In order to infer a cutoff from the data, we inspect the histogram: We visually set a cutoff line man_threshold to the left of the histogram peak in order not to exclude too many genes.

man_threshold <- 2

hist_res <- hist(palmieri_medians, 100, col = "cornsilk", freq = FALSE, 
            main = "Histogram of the median intensities",
            border = "antiquewhite4",
            xlab = "Median intensities")

abline(v = man_threshold, col = "coral4", lwd = 2)

#Transcripts that do not have intensities larger than the threshold in at least as many arrays as the smallest experimental group are excluded.

```
```{r}
#In order to do so, we first have to get a list with the number of samples (=arrays) (no_of_samples) in the experimental groups:

no_of_samples <- 
  table(paste0(pData(palmieri_eset_norm)$Factor.Value.disease., "_", 
                  pData(palmieri_eset_norm)$Factor.Value.cell.type.))
no_of_samples 

```

```{r}

#We now filter out all transcripts that do not have intensities greater than the threshold in at least as many arrays as the smallest experimental group

samples_cutoff <- min(no_of_samples)

idx_man_threshold <- apply(Biobase::exprs(palmieri_eset_norm), 1,
                           function(x){
                          sum(x > man_threshold) >= samples_cutoff})
                          table(idx_man_threshold)
                          
palmieri_manfiltered <- subset(palmieri_eset_norm, idx_man_threshold)
```

```{r}

#Before we continue with the linear models for microarrays and differential expression, we first add “feature data”, i.e. annotation information to the transcript cluster identifiers stored in the featureData of our ExpressionSet:

#stripped_feature_names <- sub("_st$", "", featureNames(palmieri_manfiltered))
#anno_palmieri <- AnnotationDbi::select(hugene10sttranscriptcluster.db,

library(org.Hs.eg.db)
library(hgu133plus2.db)
library(hgu95av2.db)
library(pd.hta.2.0)
library(AnnotationDbi)
library(hta20transcriptcluster.db)

anno_palmieri <- AnnotationDbi::select(hta20transcriptcluster.db,
                                  keys = (featureNames(palmieri_manfiltered)),
                                  columns = c("SYMBOL", "GENENAME"),
                                  keytype = "PROBEID")

anno_palmieri <- subset(anno_palmieri, !is.na(SYMBOL))
```
```{r}

#Many transcript-cluster identifiers will map to multiple gene symbols, i.e.  they can’t be unambigously assigned.

anno_grouped <- group_by(anno_palmieri, PROBEID)
anno_summarized <- 
  dplyr::summarize(anno_grouped, no_of_matches = n_distinct(SYMBOL))

head(anno_summarized)


anno_filtered <- filter(anno_summarized, no_of_matches > 1)

head(anno_filtered)

probe_stats <- anno_filtered 

nrow(probe_stats)
```

```{r}
ids_to_exlude <- (featureNames(palmieri_manfiltered) %in% probe_stats$PROBEID)

table(ids_to_exlude)

palmieri_final <- subset(palmieri_manfiltered, !ids_to_exlude)

validObject(palmieri_final)

head(anno_palmieri)
```

```{r}
fData(palmieri_final)$PROBEID <- rownames(fData(palmieri_final))

fData(palmieri_final) <- left_join(fData(palmieri_final), anno_palmieri)

# restore rownames after left_join
rownames(fData(palmieri_final)) <- fData(palmieri_final)$PROBEID 
    
validObject(palmieri_final)

```

```{r}
#linear model for the data 

individual <- as.character(Biobase::pData(palmieri_final)$Characteristics.individual.)

tissue <- str_replace_all(Biobase::pData(palmieri_final)$Factor.Value.cell.type.,
                  " ", "_")

tissue <- ifelse(tissue == "colorectal_cancer_stem-like_metastatic_cells_(mCCSCs)","mCCSC",
            ifelse(tissue == "colorectal_cancer_stem-like_circulating_cells_(CTCs)","CTC", "CCSC"))


#disease <- str_replace_all(Biobase::pData(palmieri_final)$Factor.Value.disease.,
 #                 " ", "_")
disease <- 
  ifelse(str_detect(Biobase::pData(palmieri_final)$Factor.Value.disease., 
                    "metastatic colorectal cancer"), "mCC", "CC")
```


```{r}
i_CD <- factor(individual[disease == "mCC"])
design_palmieri_CD <- model.matrix(~ 0 + tissue[disease == "mCC"] + i_CD)
colnames(design_palmieri_CD) <- c("mCCSC","CTC", "CCSC")
rownames(design_palmieri_CD) <- i_CD 

i_UC <- individual[disease == "CC"]
design_palmieri_UC <- model.matrix(~ 0 + tissue[disease == "CC"] + i_UC )
colnames(design_palmieri_UC) <- c("mCCSC","CTC", "CCSC", "other")
rownames(design_palmieri_UC) <- i_UC 
```

